facilitator/src/transport/s3.rs

```diff
@@ -741,103 +690,123 @@ mod tests {
             key: "".into(),
         };
 
-        let client_provider = Box::new(
-            |region: &Region, credentials_provider: aws_credentials::Provider| {
-                Ok(S3Client::new_with(
-                    // Failed GetObject request
-                    MockRequestDispatcher::with_status(404)
-                        .with_request_checker(is_get_object_request),
-                    credentials_provider,
-                    region.clone(),
-                ))
-            },
-        );
-        let mut transport = S3Transport::new_with_client(
-            s3_path.clone(),
-            aws_credentials::Provider::new_mock(),
-            client_provider,
-            &logger,
-        );
+        // Failed GetObject request.
+        {
+            let mocked_get_object = mock_get_object_request().with_status(404).create();
 
-        let ret = transport.get(TEST_KEY, "trace-id");
-        assert!(ret.is_err(), "unexpected return value {:?}", ret.err());
+            let mut transport = S3Transport::new_with_client(
+                s3_path.clone(),
+                aws_credentials::Provider::new_mock(),
+                Box::new(
+                    |_: &Region, credentials_provider: aws_credentials::Provider| {
+                        Ok(S3Client::new_with(
+                            HttpClient::new().expect("Could not create HTTP client"),
+                            credentials_provider,
+                            Region::Custom {
+                                name: TEST_REGION.into(),
+                                endpoint: mockito::server_url(),
+                            },
+                        ))
+                    },
+                ),
+                &logger,
+            );
 
-        let mut transport = S3Transport::new_with_client(
-            s3_path.clone(),
-            aws_credentials::Provider::new_mock(),
-            Box::new(
-                |region: &Region, credentials_provider: aws_credentials::Provider| {
-                    Ok(S3Client::new_with(
-                        // Successful GetObject request
-                        MockRequestDispatcher::with_status(200)
-                            .with_request_checker(is_get_object_request)
-                            .with_body("fake-content"),
-                        credentials_provider,
-                        region.clone(),
-                    ))
-                },
-            ),
-            &logger,
-        );
+            let ret = transport.get(TEST_KEY, "trace-id");
+            assert!(ret.is_err(), "unexpected return value {:?}", ret.err());
 
-        let mut reader = transport
-            .get(TEST_KEY, "trace-id")
-            .expect("unexpected error getting reader");
-        let mut content = Vec::new();
-        reader.read_to_end(&mut content).expect("failed to read");
-        assert_eq!(Vec::from("fake-content"), content);
+            mocked_get_object.assert();
+            mockito::reset();
```

Comment:
Are these `reset()` calls needed because we have three test cases in one function? I'd prefer to make this into three distinct tests that are less brittle.

Reply:
Yep, that's why I was making the `reset()` calls. I agree that this would be better stated as three separate test cases -- done.

Doing this also surfaced a bug (now fixed) -- mockito doesn't stop you from calling `create()` twice on the same mock, which effectively causes Mockito to register the mock twice in its internal state; evidently in this case, the mock is not cleaned up correctly by the normal drop-based cleanup (looks like one of the registrations will end up being left behind, based on a quick glance at Mockito's source). This causes some of Mockito's state from one test case to leak into later test cases. One of the mocks in `multipart_upload` was created twice, and this refactoring made the bug apparent.